### ISSUE TITLE
eventBus publisher should have a publish method

### ISSUE THREAD

**Issue by seanjburns** [eclipse/vert.x#1468]
It was slightly confusing when i used the eventBus to create publisher.

I had code that looked like this
`vertx.eventBus().publish("endpoint","Hello World");`
and wanted to use a publisher instead
`vertx.eventBus().publisher("endpoint")`

but the publisher did not have the "publish" method (the write method does the correct function though).

**Comment by EmadAlblueshi** [comment]
@seanjburns Hi, I hope I got what you need

```javascript
DeliveryOptions options = new DeliveryOptions();
options.addHeader("action", "findAll");

MessageProducer<JsonObject> publisherMessage = eventBus.<JsonObject>publisher("io.vertx.news", options);
MessageProducer<JsonObject> senderMessage = eventBus.<JsonObject>sender("io.vertx.news", options);

JsonObject message = new JsonObject();

publisherMessage.<JsonObject>send(message, reply -> {
// publish
});

senderMessage.<JsonObject>send(message, reply -> {
// send
});
```

**Comment by seanjburns** [comment]
^^ You see, I'm not the only one easily "fooled" @EmadAlblueshi just pasted a "solution" thats does not "publish" as expected.

@EmadAlblueshi this function
`publisherMessage.<JsonObject>send(message,` reply -> {
// publish <==== THIS DOES NOT PUBLISH
});`

Does not actually "publish" to all consumers, it "sends" to a single consumer. You actually have to use the write method like this

`publisherMessage.<JsonObject>write(message,` reply -> {
// publish
});`

FYI this is just a usability issue that is all.

**Comment by EmadAlblueshi** [comment]
@seanjburns you are absolutely correct. I have just discovered that "write" method will do the publishing instead of sending. From usability perspective I think it should be "publish" to vanish such a confusion

+1

**Comment by seanjburns** [comment]
https://groups.google.com/d/msg/vertx/b_0SuxHEQ40/nFDNMpENAQAJ

**Comment by vietj** [comment]
I think having a `send` method was a design mistake,  adding a `publish` method would be even more confusing and people can still do `eventBus.publish` and `eventBus.send`

**Comment by seanjburns** [comment]
Maybe for vertx 4.0 (3.4?), We can have 2 interfaces that extend MessageProducer. Then the publisher and sender methods can return the respective interfaces.

**Comment by vietj** [comment]
@seanjburns I was rather thinking deprecating the `send` method and find a new name that works for both unless we want to keep only the `write`

**Comment by barbarosalp** [comment]
Why don't you use `vertx.eventBus().publish(eventBusAddressForPublish, event);` ?
eventBus is a cached object already.

**Comment by barbarosalp** [comment]
If you don't want to provide the address all the time, just create a publisher with an address:
- to publish: use its `write` method if you want to **publish** a message to all consumers.
- to send:  use its `send` method. (If there are multiple consumers the message will be sent round-robin)
